src/ipips/ipip-0402.md

```diff
@@ -0,0 +1,337 @@
+
+As long the client was written in a trustless manner, and follows ring and was discarding
+unexpected blocks, this will be a backward-compatible change.
+
+#### CAR format with `entity-bytes` and `dag-scope` parameters
+
+These parameters are opt-in, which means no breaking changes.
+
+Gateways ignore unknown URL parameters. A client sending them to a
+gateway that does not implement this IPIP will get all blocks for the requested
+DAG.
+
+<!-- TODO
+mention feature detection via OPTIONS -- a separate IPIP?
+OR suggest executing test request and client-side detection the first time a gateway is used.
+-->
+
+### Security
+
+This IPIP allows clients to narrow down the amount of data returned as a CAR,
+and introduces a need for defensive programming when the feature set of the
+remote gateway is unknown.
+
+To avoid denial of service, and resource starvation, clients should probe if
+the gateway supports features described in this IPIP before requesting data, to
+avoid fetching big DAGs when only a small subset of blocks is expected.
+
+Following the robustness principle, invalid, duplicate or unexpected blocks
+should be discarded.
+
+### Alternatives
+
+Below are alternate designs that were considered, but decided to be out of scope for this IPIP.
+
+#### Arbitrary IPLD Selectors
+
+Passing arbitrary selectors was rejected due to the implementation complexity,
+risks, and weak value proposition, as [discussed during IPFS Thing 2022](https://github.com/ipfs/specs/issues/348#issuecomment-1326869509)
+
+#### Additional "Web" Scope
+
+A request for
+`/ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze/wiki/?format=car&dag-scope=entity`
+returns all blocks required for enumeration of the big HAMT `/wiki` directory,
+and then an additional request for `index.html` needs to be issued.
+
+Website hosting use case could be made more efficient if gateway returned a CAR
+with `index.html` instead of all blocks for directory enumeration. The server
+already did the work: it knows the entity is a directory, already parsed it, it
+knows it has child entity named `index.html`, and everyone would pay a lower cost due
+to lower number of blocks being returned in a single round-trip, instead of two.
+
+Rhea/Saturn projects requested this to be out of scope for now, but this "web"
+entity scope could be added in the future, as a follow-up optimiziation IPIP.
+
+#### Requesting specific DAG depth
+
+Blindly requesting specific DAG depth did not translate to any type of
+requests web gateways like `ipfs.io` or one in Brave browser have to handle.
+
+It is impossible to know if some entity on a sub-path is a file or a directory,
+without sending a probe for the root block, which introduces one round-trip overhead
+per entity.
+
+This problem is not present in the case of `dag-scope=entity`, which shifts the
+decision to the server, and allows for fetching unknown UnixFS entity with a
+single request.
+
+## Test fixtures
```

Comment:
@guanzo @tchardin would appreciate feedback on below test scenarios. (cc @laurentsenta  @galargh  for visibility).

The intention here is to define the bare minimum we need in gateway-conformance to ensure independent implementations of IPFS do the right thing. Anything to add? Remove/simplify?

Reply:
WIP PR for car-scope https://github.com/ipfs/gateway-conformance/pull/49/files

Reply:
cc @aschmahmann as he works on `boxo/gateway` implementation (https://github.com/ipfs/boxo/pull/303)